providers/common/provider_util.c

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+#include "internal/provider_util.h"
+
+void ossl_prov_cipher_free(PROV_CIPHER *pc)
+{
+    EVP_CIPHER_free(pc->alloc_cipher);
+    pc->alloc_cipher = NULL;
+}
+
+int ossl_prov_cipher_dup(PROV_CIPHER *dst, const PROV_CIPHER *src)
+{
+    if (src->alloc_cipher != NULL && !EVP_CIPHER_up_ref(src->alloc_cipher))
+        return 0;
+    dst->engine = src->engine;
+    dst->cipher = src->cipher;
+    dst->alloc_cipher = src->alloc_cipher;
+    return 1;
+}
+
+static const OSSL_PARAM *load_common(const OSSL_PARAM params[],
+                                     const char **propquery, ENGINE **engine,
+                                     const char *pname)
+{
+    const OSSL_PARAM *p;
+
+    *propquery = NULL;
+    p = OSSL_PARAM_locate_const(params, "properties");
+    if (p != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return NULL;
+        *propquery = p->data;
+    }
+
+    *engine = NULL;
+    /* TODO(3.0) legacy stuff, to be removed */
```

Comment:
We can't remove ENGINE support at 3.0, that would be against policy. Keep the TODO, lose "3.0"